Missing Android consumer proguard rules

**Describe the bug**
We started using shrinking/obfuscation for our java code by setting `minifyEnabled true` in our `build.gradle`. We build release builds for testing with the command shown [here](https://github.com/wix/Detox/blob/master/docs/Introduction.Android.md#2-apply-detox-configuration) via `./gradlew assembleRelease assembleAndroidTest -DtestBuildType=release`. In your docs you mention adding your proguard file shown [here](https://github.com/wix/Detox/blob/master/detox/android/detox/proguard-rules-app.pro), but that doesn't seem to be enough to test on minified builds, and either our the app will crash in tests or Detox will log errors and tests will fail. 

First, we were experiencing immediate crashes when running tests from Exceptions relating to missing classes. There is an existing issue with Android Instrumentation testing in general when minified, so Slack made a plugin called [Keeper](https://slackhq.github.io/keeper/) to address this for now. The short explanation is that Detox relies on some dependencies in the main app apk like Kotlin and React-Native that the test apk is then not able to find at runtime. They are being removed from the main apk during the shrinking process because the main app is not using those classes. This plugin analyzes test code and adds Proguard rules for the main app for classes the test apk relies on. That's a separate issue, but just noting as it's not mentioned in your docs as potential watch-out.

That fixed some crashes, but your library also uses reflection in a number of your classes, and the compiler cannot know what classes you're trying to use dynamically, so they could get removed or obfuscated. We then need additional proguard rules that don't seem to be in your `proguard-rules-app.pro` file. For example, we were using react-native `0.63.2` and Detox `17.6.1`, and we see several errors like this when running tests on minified builds: 

```
E Detox   : Can't set up RN UIModule listener
E Detox   : org.joor.ReflectException: java.lang.NoSuchFieldException: mOperationsQueue
E Detox   : 	at org.joor.Reflect.field0(Reflect.java:310)
E Detox   : 	at org.joor.Reflect.field(Reflect.java:282)
E Detox   : 	at com.wix.detox.reactnative.idlingresources.UIModuleIdlingResource.isIdleNow(UIModuleIdlingResource.java:84)
E Detox   : 	at androidx.test.espresso.base.IdlingResourceRegistry.allResourcesAreIdle(IdlingResourceRegistry.java:268)
E Detox   : 	at androidx.test.espresso.base.IdlingResourceRegistry$6.isIdleNow(IdlingResourceRegistry.java:302)
E Detox   : 	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:453)
E Detox   : 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:275)
E Detox   : 	at androidx.test.espresso.ViewInteraction$2.call(ViewInteraction.java:272)
E Detox   : 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E Detox   : 	at android.os.Handler.handleCallback(Handler.java:790)
E Detox   : 	at android.os.Handler.dispatchMessage(Handler.java:99)
E Detox   : 	at android.os.Looper.loop(Looper.java:164)
E Detox   : 	at android.app.ActivityThread.main(ActivityThread.java:6494)
E Detox   : 	at java.lang.reflect.Method.invoke(Native Method)
E Detox   : 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
E Detox   : 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
E Detox   : Caused by: java.lang.NoSuchFieldException: mOperationsQueue
E Detox   : 	at java.lang.Class.getField(Class.java:1601)
E Detox   : 	at org.joor.Reflect.field0(Reflect.java:295)
E Detox   : 	... 15 more
```

I was able to fix some of these just by adding Proguard rules to our main app to just keep all of certain packages as I didn't know all the classes you might use reflection on.

```
-keep class com.facebook.react.modules.** { *; }
-keep class com.facebook.react.animated.** { *; }
-keep class com.facebook.react.uimanager.** { *; }
```

However, today we tried upgrading to `17.10.3` as there are new features and that led to a new crash:

```
E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
    Process: to.hover.android.app, PID: 26450
    org.joor.ReflectException: org.joor.ReflectException: java.lang.NoSuchFieldException: mTasks
        at org.joor.Reflect.field(Reflect.java:286)
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.pendingTasks(SerialExecutorReflected.kt:14)
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.hasPendingTasks(SerialExecutorReflected.kt:9)
        at com.wix.detox.reactnative.idlingresources.AsyncStorageIdlingResource$idleCheckTaskImpl$1.run(AsyncStorageIdlingResource.kt:39)
        at com.reactnativecommunity.asyncstorage.AsyncStorageModule$g$a.run(AsyncStorageModule.java:1)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: org.joor.ReflectException: java.lang.NoSuchFieldException: mTasks
        at org.joor.Reflect.field0(Reflect.java:310)
        at org.joor.Reflect.field(Reflect.java:282)
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.pendingTasks(SerialExecutorReflected.kt:14) 
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.hasPendingTasks(SerialExecutorReflected.kt:9) 
        at com.wix.detox.reactnative.idlingresources.AsyncStorageIdlingResource$idleCheckTaskImpl$1.run(AsyncStorageIdlingResource.kt:39) 
        at com.reactnativecommunity.asyncstorage.AsyncStorageModule$g$a.run(AsyncStorageModule.java:1) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
     Caused by: java.lang.NoSuchFieldException: mTasks
        at java.lang.Class.getField(Class.java:1601)
        at org.joor.Reflect.field0(Reflect.java:295)
        at org.joor.Reflect.field(Reflect.java:282) 
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.pendingTasks(SerialExecutorReflected.kt:14) 
        at com.wix.detox.reactnative.idlingresources.SerialExecutorReflected.hasPendingTasks(SerialExecutorReflected.kt:9) 
        at com.wix.detox.reactnative.idlingresources.AsyncStorageIdlingResource$idleCheckTaskImpl$1.run(AsyncStorageIdlingResource.kt:39) 
        at com.reactnativecommunity.asyncstorage.AsyncStorageModule$g$a.run(AsyncStorageModule.java:1) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
```

It appears that another class `AsyncStorageIdlingResource` was added that uses reflection on the `com.reactnativecommunity.asyncstorage.AsyncStorageModule` class from another library we use.

**My ask then is can you provide the necessary proguard rules for any case in which you use reflection?** It's important for us to be able to test on minified builds, and it's difficult and unsustainable for us to look through your code and figure our these rules ourselves and things could still break when you add something new with reflection in between upgrades, like happened with us. These rules can just be included in your `proguard-rules-app.pro` file or at least suggestions for rules in your docs. Thank you.

--- a/detox/android/detox/proguard-rules-app.pro
+++ b/detox/android/detox/proguard-rules-app.pro
@@ -1,1 +1,9 @@
+-keepattributes InnerClasses, Exceptions
+-keep class com.facebook.react.modules.** { *; }
+-keep class com.facebook.react.uimanager.** { *; }
+-keep class com.facebook.react.animated.** { *; }
+-keep class com.facebook.react.ReactApplication { *; }
+-keep class com.facebook.react.ReactNativeHost { *; }
 -keep class com.facebook.react.ReactInstanceManager { *; }
+-keep class com.facebook.react.ReactInstanceManager** { *; }
+-keep class com.reactnativecommunity.asyncstorage.** { *; }

--- a/detox/android/detox/proguard-rules.pro
+++ b/detox/android/detox/proguard-rules.pro
@@ -13,12 +13,6 @@
 -dontnote org.hamcrest.**
 -dontnote com.facebook.**
 
-# Detox dynamic access to declared methods of RN's timing module
-
--keep class com.facebook.react.modules.core.TimingModule { *; } # RN >= .63
--keep class com.facebook.react.modules.core.Timing { *; } # RN <= .62
--dontnote com.wix.detox.reactnative.idlingresources.timers.**
-
 # Detox profiler (optional)
 
 -keep class com.wix.detoxprofiler.** { *; }

--- a/detox/android/detox/src/main/java/com/wix/detox/reactnative/helpers/RNHelpers.kt
+++ b/detox/android/detox/src/main/java/com/wix/detox/reactnative/helpers/RNHelpers.kt
@@ -12,9 +12,9 @@ object RNHelpers {
             val moduleClass = Class.forName(className) as Class<NativeModule>
 
             if (reactContext.hasNativeModule(moduleClass)) {
-                Log.d(LOG_TAG, "Native module not resolved: no registered module")
                 reactContext.getNativeModule(moduleClass)
             } else {
+                Log.d(LOG_TAG, "Native module not resolved: no registered module")
                 null
             }
         } catch (ex: Exception) {

--- a/detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResource.kt
+++ b/detox/android/detox/src/main/java/com/wix/detox/reactnative/idlingresources/AsyncStorageIdlingResource.kt
@@ -30,6 +30,9 @@ open class AsyncStorageIdlingResource
         sexecutorReflectedGenFn: SExecutorReflectedGenFnType = defaultSExecutorReflectedGenFn)
     : IdlingResource {
 
+    open val logTag: String
+        get() = LOG_TAG
+
     private val moduleReflected = ModuleReflected(module, sexecutorReflectedGenFn)
     private var callback: IdlingResource.ResourceCallback? = null
     private var idleCheckTask: Runnable? = null
@@ -50,7 +53,7 @@ open class AsyncStorageIdlingResource
     override fun isIdleNow(): Boolean =
         checkIdle().also { idle ->
             if (!idle) {
-                Log.d(LOG_TAG, "Async-storage is busy!")
+                Log.d(logTag, "Async-storage is busy!")
                 enqueueIdleCheckTask()
             }
         }
@@ -106,4 +109,7 @@ open class AsyncStorageIdlingResource
     }
 }
 
-class AsyncStorageIdlingResourceLegacy(module: NativeModule): AsyncStorageIdlingResource(module)
+class AsyncStorageIdlingResourceLegacy(module: NativeModule): AsyncStorageIdlingResource(module) {
+    override val logTag: String
+        get() = super.logTag + "Legacy"
+}

--- a/detox/test/android/app/proguard-rules.pro
+++ b/detox/test/android/app/proguard-rules.pro
@@ -16,6 +16,7 @@
 -dontnote okio.**
 
 # Do not strip any method/class that is annotated with @DoNotStrip
+# This should really come from React Native itself. See here: https://github.com/react-native-community/upgrade-support/issues/31
 -keep @com.facebook.jni.annotations.DoNotStrip class *
 -keep class * {
     @com.facebook.proguard.annotations.DoNotStrip *;

--- a/docs/Introduction.Android.md
+++ b/docs/Introduction.Android.md
@@ -210,9 +210,9 @@ For full details, refer to [Android's security-config guide](https://developer.a
 
 
 
-### 7. Proguard (Minification)
+### 7. Proguard (Minification, Obfuscation)
 
-In apps running [minification using Proguard](https://developer.android.com/studio/build/shrink-code), in order for Detox to work well on release builds, please enable some Detox proguard-configuration rules by applying the custom configuration file on top of your own. Typically, this is defined using the `proguardFiles` statement in the minification-enabled build-type in your `android/app/build.gradle`:
+In apps running [minification using Proguard](https://developer.android.com/studio/build/shrink-code), in order for Detox to work well on release builds, please enable some Detox proguard-configuration rules by applying the custom configuration file on top of your own. Typically, this is defined using the `proguardFiles` statement in the minification-enabled build-type in your `app/build.gradle`:
 
 ```groovy
     buildTypes {
@@ -229,7 +229,33 @@ In apps running [minification using Proguard](https://developer.android.com/stud
 
 ```
 
+:warning: **Note:** In order for Detox to be able to work properly, in `proguard-rules-app.pro`, it effectively declares rules that retain most of React-Native's code (i.e. keep it unminified, unobfuscated) in your **production** APK. Though generally speaking, this should not be an issue (as React-Native is an open-source project), there are ways around that, if it bothers you. For example, running your E2E over a build-type specifically designed to run E2E tests using Detox would do the trick -- roughly, like so (in `app/build.gradle`):
 
+```groovy
+    buildTypes {
+        release {
+            minifyEnabled true
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+
+            signingConfig signingConfigs.release
+        }
+        releaseE2E {
+            initWith release
+            setMatchingFallbacks('release')
+
+            proguardFile "${rootProject.projectDir}/../node_modules/detox/android/detox/proguard-rules-app.pro"
+        }
+    }
+```
+
+Here we utilize Gradle's `initWith` to easily define `releaseE2E` in a way that is identical to the `release` build-type, with the exception of considering Detox' `proguard-rules-app.pro` in the minification process.
+
+Following the example, you would then have to build your app using `gradlew assembleReleaseE2E` rather than `gradlew assembleRelease` before running Detox, and instruct Detox (i.e. via `binaryPath` in the Detox configuration file) to use the APK resulted specifically by *that* Gradle target (e.g. in `app/build/apk/releaseE2E/app-releaseE2E.apk` instead of the equivalent `app/build/apk/release/app-release.apk`).
+
+> Note: if you app contains flavours -- that makes things a bit trickier, but the approach can generally be adjusted to support that as well.
+
+**Last but not least:** If you're having issue with Detox' Proguard rules, please report them [here](https://github.com/wix/Detox/issues/new/choose).
+A special thanks to [GEllickson-Hover](https://github.com/GEllickson-Hover) for reporting issues related to obfuscation in [#2431](https://github.com/wix/Detox/issues/2431).
 
 ### 8. Test Butler Support (Optional)
 

--- a/examples/demo-react-native/android/app/build.gradle
+++ b/examples/demo-react-native/android/app/build.gradle
@@ -69,6 +69,8 @@ dependencies {
 
     // noinspection GradleDynamicVersion
     implementation 'com.facebook.react:react-native:+'
+    // noinspection GradleDynamicVersion
+    implementation(project(path: ':AsyncStorage')) // Note: not required unless effectively used by your app
 
     androidTestImplementation('com.wix:detox:+')
 }

--- a/examples/demo-react-native/android/app/proguard-rules.pro
+++ b/examples/demo-react-native/android/app/proguard-rules.pro
@@ -3,8 +3,6 @@
 # the lean configuration file.
 ###
 
--dontobfuscate
-
 -dontnote android.net.**
 -dontnote org.apache.**
 
@@ -16,6 +14,7 @@
 -dontnote okio.**
 
 # Do not strip any method/class that is annotated with @DoNotStrip
+# This should really come from React Native itself. See here: https://github.com/react-native-community/upgrade-support/issues/31
 -keep @com.facebook.jni.annotations.DoNotStrip class *
 -keep class * {
     @com.facebook.proguard.annotations.DoNotStrip *;

--- a/examples/demo-react-native/android/app/src/main/java/com/example/MainApplication.java
+++ b/examples/demo-react-native/android/app/src/main/java/com/example/MainApplication.java
@@ -7,6 +7,7 @@ import com.facebook.react.ReactNativeHost;
 import com.facebook.react.ReactPackage;
 import com.facebook.react.shell.MainReactPackage;
 import com.facebook.soloader.SoLoader;
+import com.reactnativecommunity.asyncstorage.AsyncStoragePackage;
 
 import java.util.Arrays;
 import java.util.List;
@@ -21,7 +22,8 @@ public class MainApplication extends Application implements ReactApplication {
         @Override
         protected List<ReactPackage> getPackages() {
             return Arrays.<ReactPackage>asList(
-                    new MainReactPackage()
+                    new MainReactPackage(),
+                    new AsyncStoragePackage() // Note: Not required unless effectively used by your app
             );
         }
     };

--- a/examples/demo-react-native/android/build.gradle
+++ b/examples/demo-react-native/android/build.gradle
@@ -5,7 +5,7 @@ buildscript {
     println "[$project] Resorted to Android Gradle-plugin version $androidGradlePluginVersion"
 
     ext {
-        kotlinVersion = '1.3.41'
+        kotlinVersion = '1.4.21'
         buildToolsVersion = '29.0.3'
         compileSdkVersion = 29
         targetSdkVersion = 29

--- a/examples/demo-react-native/android/gradle.properties
+++ b/examples/demo-react-native/android/gradle.properties
@@ -16,3 +16,7 @@ org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryErro
 # This option should only be used with decoupled projects. More details, visit
 # http://www.gradle.org/docs/current/userguide/multi_project_builds.html#sec:decoupled_projects
 # org.gradle.parallel=true
+
+android.useAndroidX=true
+android.enableJetifier=true
+

--- a/examples/demo-react-native/android/settings.gradle
+++ b/examples/demo-react-native/android/settings.gradle
@@ -1,3 +1,6 @@
 rootProject.name = 'DetoxRNExample'
 
 include ':app'
+
+include ':AsyncStorage'
+project(':AsyncStorage').projectDir = new File(rootProject.projectDir, '../node_modules/@react-native-community/async-storage/android')

--- a/examples/demo-react-native/package.json
+++ b/examples/demo-react-native/package.json
@@ -19,7 +19,8 @@
   },
   "dependencies": {
     "react": "16.11.x",
-    "react-native": "0.62.x"
+    "react-native": "0.62.x",
+    "@react-native-community/async-storage": "^1.12.0"
   },
   "devDependencies": {
     "detox": "^17.14.3",

